Legalization

hibenj opened the pull request:
## Description

Adds the following .hpp and .cpp files:
- check_planarity: For a given network with a rank view, check for planarity
- virtual_pi_network: stores additional PIs, which are duplicates of other pis. These are called virtual PIs and are stored with a map to their origin PI
- extended_rank_view: the extended rank_view allows for a more easy modification of ranks and also adds another constructor from an array of nodes. (This might also be more error prone)
- node_suplication_planarization: Constucts an H-graph (as in "Fabricatable Interconnect and
Molecular QCA Circuits") in order to duplicate nodes on a given PO order. The output is a extended_rank_view of a virtual_pi_network

Fixes # (issue)

## Checklist:

<!---
This checklist serves as a reminder of a couple of things that ensure your pull request will be merged swiftly.
-->

- [ ] The pull request only contains commits that are related to it.
- [ ] I have added appropriate tests and documentation.
- [ ] I have added a changelog entry.
- [ ] I have created/adjusted the Python bindings for any new or updated functionality.
- [ ] I have made sure that all CI jobs on GitHub pass.
- [ ] The pull request introduces no new warnings and follows the project's style guidelines.

marcelwa commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,693 @@
+    /**
+     * @brief Reimplementation of is_pi(). Check if a given node is a primary input (PI) in the virtual_pi_network.
+     *
+     * A node is considered a PI if it exists in the virtual_inputs vector or the inputs vector of the _storage object.
```
This is another good example of *how* vs. *what*. I would reimagine this docstring as follows:
```suggestion
     * A node is considered a PI if it is a virtual PI created via `create_virtual_pi()` or a real PI created in the original network via `create_pi()`.
```

marcelwa commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,693 @@
+     * @param n The node to check.
+     * @return True if the node is a virtual PI, false otherwise.
+     */
+    [[nodiscard]] bool is_pi_virtual(node const& n) const
```
```suggestion
    [[nodiscard]] bool is_virtual_pi(node const& n) const
```
Because we cannot limit the input to PIs, this function checks for both the node being a PI and it being real, thus the proposed name is more accurate.

marcelwa commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,693 @@
+     */
+    [[nodiscard]] bool is_pi_virtual(node const& n) const
+    {
+        return std::find(virtual_inputs->begin(), virtual_inputs->end(), n) != virtual_inputs->end();
```
```suggestion
        return std::find(virtual_inputs->cbegin(), virtual_inputs->cend(), n) != virtual_inputs->cend();
```

marcelwa commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,693 @@
+     * @param n The node to check.
+     * @return True if the node is a real PI, false otherwise.
+     */
+    [[nodiscard]] bool is_pi_real(node const& n) const
+    {
+        return std::find(_storage->inputs.begin(), _storage->inputs.end(), n) != _storage->inputs.end();
+    }
```
See above: name and `cost`ness

marcelwa commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,693 @@
+     * @param n The node to check.
+     * @return True if the node is a const input, false otherwise.
+     */
+    [[nodiscard]] bool is_ci(node const& n) const
```
Comments from above apply to all further functions as well

marcelwa commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,693 @@
+        if (it != map.end())
+            return it->second;
+
+        throw std::runtime_error("Error: node is not a virtual pi");
```
We usually don't work with exceptions in this case as they have a large impact on performance. Rather add an `assert` to ensure that the given node is indeed a PI.

marcelwa commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,693 @@
+    {
+        auto it = std::find_if(map.begin(), map.end(), [v_pi](const auto& pair) { return pair.first == v_pi; });
+
+        if (it != map.end())
+            return it->second;
```
```suggestion
        if (it != map.end())
        {
            return it->second;
        }
```
Always wrap your statements. Not doing it, is bug-prone and has led to critical bugs in the real world.

marcelwa reviewed (changes requested):
Many, many thanks for this great addition. While a complete code review, will take some more time, I have assembled some more high-level comments to the `virtual_pi_network` in particular. I hope you will find them useful. Most comments also apply to other files in a similar manner (like `cost`ness, includes, comments, docstrings, etc.). Please try to actively look out for these instances address them in other files already. This will greatly speed up the remaining review process. Once again, thank you for your hard work!

Drewniok commented:
@hibenj I have noticed that the docstring generator is still encountering issues at your end, even though it was supposed to be fixed. Could you please check in your fork if the settings for GitHub Actions are configured correctly? Make sure that GitHub Actions have the necessary read and write access to the repository.

hibenj commented:
I refactored the virtual_pi_network.hpp. Also i added a virtual_miter, which_checks for virtual_pi_networks as well. maybe the file can just be named miter.hpp, since there is no such file in fiction yet.
I left in the comments for the special constructor, since the construction of the network only works like this because of the algorithm used in the node_duplication_planarization.hpp. I think this should be moved to this file again, since this constructor will not be used in any other case.
Due to the refactoring of this file some other new files don't work, but lets focus on this first. I will fix the other issues in the meantime.

marcelwa commented on test/algorithms/network_transformation/node_duplication_planarization.cpp:
```diff
@@ -0,0 +1,117 @@
+    CHECK(*cec_m == 1);
+}
+
+TEST_CASE("Aig Test", "[node-duplication-planarization]")
```
```suggestion
TEST_CASE("AIG test", "[node-duplication-planarization]")
```
Please replace "test" with a descriptor of what is exactly tested.

marcelwa commented on include/fiction/algorithms/properties/check_planarity.hpp:
```diff
@@ -0,0 +1,104 @@
+ * @throw std::runtime_error if the network is not balanced.
+ */
+template <typename Ntk>
+bool check_planarity(const Ntk& ntk)
```
```suggestion
[[nodiscard]] bool check_planarity(const Ntk& ntk)
```

marcelwa commented on include/fiction/algorithms/properties/check_planarity.hpp:
```diff
@@ -0,0 +1,104 @@
+template <typename Ntk>
+bool check_planarity(const Ntk& ntk)
+{
+    static_assert(mockturtle::has_get_node_v<Ntk>, "Ntk does not have get_node trait");
+    static_assert(mockturtle::has_foreach_fanin_v<Ntk>, "Ntk does not have foreach_fanin trait");
+    static_assert(mockturtle::has_depth_v<Ntk>, "Ntk does not have depth trait");
+    static_assert(mockturtle::has_rank_position_v<Ntk>, "Ntk does not have rank_position trait");
+    static_assert(mockturtle::has_foreach_node_in_rank_v<Ntk>, "Ntk does not have foreach_node_in_rank trait");
```
We usually warn via "Ntk does not implement the ... function"

marcelwa commented on include/fiction/algorithms/properties/check_planarity.hpp:
```diff
@@ -0,0 +1,104 @@
+ * @tparam Ntk Logic network type.
+ * @param ntk The logic network to check for planarity.
+ * @return `true` if the network is planar, `false` otherwise.
+ * @throw std::runtime_error if the network is not balanced.
```
```suggestion
 * @throw `std::runtime_error` if the network is not balanced.
```

marcelwa commented on include/fiction/algorithms/properties/check_planarity.hpp:
```diff
@@ -0,0 +1,104 @@
+/**
+ * Checks if a logic network is planar. To perform this check, the network must have ranks assigned.
+ *
+ * If the network is not balanced, an exception is thrown. To balance the network, insert buffers to divide multi-level
+ * edges.
```
Why can only balanced networks be checked for planarity? Because you check level by level? Is that a true planarity check then? Should we reflect this in the function name and docstring? Something like `check_rank_planarity` or `check_rank_by_rank_planarity` (just ideas after thinking about it for two seconds; you might come up with something better).

marcelwa commented on include/fiction/algorithms/properties/check_planarity.hpp:
```diff
@@ -0,0 +1,104 @@
+    check_planarity_impl(const Ntk& ntk) : ntk(ntk) {}
+
+    /**
+     * @brief Checks if a given network is planar.
```
We usually don't use `@brief`

marcelwa commented on include/fiction/algorithms/properties/check_planarity.hpp:
```diff
@@ -0,0 +1,104 @@
+     *
+     * This function checks if the network represented by the variable `ntk` is planar.
+     * The network is planar if for any edge with starting point m and endpoint n (represented by the node ranks), there
+     * is never another edge with starting point m_ > m and endpoint n_ < n, or vice versa. When iterating through
```
You can use in-line MathJax via `\f$ ... \f$` if you want.

marcelwa commented on include/fiction/algorithms/properties/check_planarity.hpp:
```diff
@@ -0,0 +1,104 @@
+                                             return;
+                                         bound = new_bound;
+                                     });
+            if (return_false)
+                return false;
```
Please make sure to always wrap your blocks everywhere. It's error-prone not to. clang-tidy should have warned you about it in your IDE. Please make sure there are no IDE warnings left.

marcelwa commented on include/fiction/algorithms/properties/check_planarity.hpp:
```diff
@@ -0,0 +1,104 @@
+    /**
+     * @brief Checks if a given network is planar.
+     *
+     * This function checks if the network represented by the variable `ntk` is planar.
+     * The network is planar if for any edge with starting point m and endpoint n (represented by the node ranks), there
+     * is never another edge with starting point m_ > m and endpoint n_ < n, or vice versa. When iterating through
+     * the ranks of one level, the endpoints are always increasing. Therefore, only the starting points need to be
+     * checked. Thus, the highest connected starting point in the fan-in gives a border m_max for every subsequent edge.
```
This description is really nice and could be added to the API function below, which is currently a bit sparsely documented. This docstring here is only visible to developers. The one below also to users.

marcelwa commented on include/fiction/algorithms/properties/check_planarity.hpp:
```diff
@@ -0,0 +1,104 @@
+                                         ntk.foreach_fanin(n,
+                                                           [this, &bound, &new_bound, &return_false](const auto& fi)
+                                                           {
+                                                               const auto fi_n = ntk.get_node(fi);
+                                                               if (ntk.rank_position(fi_n) < bound)
+                                                               {
+                                                                   return_false = true;
+                                                               }
+                                                               new_bound = std::max(new_bound, ntk.rank_position(fi_n));
```
```suggestion
                                                               const auto fi_n = ntk.get_node(fi);
                                                               if (ntk.rank_position(fi_n) < bound)
                                                               {
                                                                   return_false = true;
                                                                   return false;  // stop iterating
                                                               }
                                                               
                                                               new_bound = std::max(new_bound, ntk.rank_position(fi_n));
                                                               
                                                               return true;  // keep iterating
```

marcelwa reviewed (changes requested):
Some comments on planarity checking. Will get to the rest as quickly as I can.

marcelwa commented on include/fiction/algorithms/verification/virtual_miter.hpp:
```diff
@@ -0,0 +1,121 @@
+ * is `nullopt`, whenever the two input networks don't match in their number of
+ * primary inputs and primary outputs.
+ */
+template <class NtkDest, class NtkSource1, class NtkSource2>
```
We usually use `typename` instead of `class` in `template`s

marcelwa commented on include/fiction/algorithms/verification/virtual_miter.hpp:
```diff
@@ -0,0 +1,121 @@
+ * primary inputs and primary outputs.
+ */
+template <class NtkDest, class NtkSource1, class NtkSource2>
+std::optional<NtkDest> miter(NtkSource1 const& ntk1, NtkSource2 const& ntk2)
```
We use west-`const` instead of east-`const`

marcelwa replied:
The file and the header guards are referring to a virtual miter. This function is just called `miter`. There seems to be a mismatch. I would prefer this function to also indicate in its name that there is more to it than `mockturtle::miter`.

marcelwa commented on include/fiction/algorithms/verification/virtual_miter.hpp:
```diff
@@ -0,0 +1,121 @@
+namespace fiction
+{
+
+/*! \brief Creates a combinational miter from two networks.
```
We use `@` instead of `\` but we don't usually use `brief`

marcelwa commented on include/fiction/algorithms/verification/virtual_miter.hpp:
```diff
@@ -0,0 +1,121 @@
+
+/*! \brief Creates a combinational miter from two networks.
+ *
+ * This method combines two networks that have the same number of primary
+ * inputs and the same number of primary outputs into a miter.  The miter
+ * has the same number of inputs and one primary output.  This output is the
+ * OR of XORs of all primary output pairs.  In other words, the miter outputs
+ * 1 for all input assignments in which the two input networks differ.
+ *
+ * All networks may have different types.  The method returns an optional, which
+ * is `nullopt`, whenever the two input networks don't match in their number of
+ * primary inputs and primary outputs.
```
This does not discuss how this function differs from `mockturtle::miter`. Please mention virtual PIs here.

marcelwa commented on include/fiction/algorithms/verification/virtual_miter.hpp:
```diff
@@ -0,0 +1,121 @@
+    std::shared_ptr<NtkSource1> ntk1_copy;
+    if constexpr (has_remove_virtual_input_nodes_v<NtkSource1>)
+    {
+        ntk1_copy = std::make_shared<NtkSource1>(ntk1);
```
Why the pointers?

marcelwa replied:
This is not a deep copy. I believe the networks have a `clone()` (?) function for that

marcelwa commented on include/fiction/algorithms/verification/virtual_miter.hpp:
This function requires a stand-alone test file that extensively tests its different use cases (datapaths)

marcelwa reviewed (changes requested):
Comments on the virtual miter function

Drewniok commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,516 @@
+{
+
+template <typename Ntk>
+struct virtual_storage
```
I would put this inside the class as ``public``.

Drewniok commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,516 @@
+     * Default constructor for the `virtual_pi_network` class.
+     * Initializes `_storage` as a shared pointer.
+     */
+    explicit virtual_pi_network() : _storage(std::make_shared<virtual_storage<Ntk>>()) {}
```
we should add ``static_assert``'s here, right?

Drewniok commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,516 @@
+     *
+     * @param ntk The network to clone into this object.
+     */
+    explicit virtual_pi_network(const Ntk& ntk) : Ntk(ntk.clone()), _storage(std::make_shared<virtual_storage<Ntk>>())
```
Shouldn't we copy the storage as well? What is the purpose of copying only the network?

hibenj replied:
For this purpose we use `clone`. This constructor makes it possible to add the functionalities of a `virtual_pi_network` on top of a network.
`TestType ntk{};

    const auto a = ntk.create_pi();
    const auto b = ntk.create_pi();

    const auto f1 = ntk.create_and(a, b);

    virtual_pi_network vpi{ntk};

    const auto c = vpi.create_virtual_pi(a);`
    
    Tbh this has no real use yet, i may also remove it

Drewniok commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,516 @@
+    using node    = typename Ntk::node;
+    using signal  = typename Ntk::signal;
+
+  public:
```
```suggestion

```

Drewniok commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,516 @@
+     */
+    template <class NtkDest, class NtkSrc>
+    std::pair<NtkDest, mockturtle::node_map<std::vector<mockturtle::signal<NtkDest>>, NtkSrc>>
+    initialize_copy_network_virtual(NtkSrc const& src)
```
can this function be private?

Drewniok commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,516 @@
+     * @return A pair of the destination network and a node map from the source to the destination network.
+     */
+    template <class NtkDest, class NtkSrc>
+    std::pair<NtkDest, mockturtle::node_map<std::vector<mockturtle::signal<NtkDest>>, NtkSrc>>
```
```suggestion
    [[nodiscard]] std::pair<NtkDest, mockturtle::node_map<std::vector<mockturtle::signal<NtkDest>>, NtkSrc>>
```

Drewniok replied:
Do we even need ``NtkDest`` since this function is only used inside this class?

Drewniok commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,516 @@
+     * To manage the connections between nodes, this also uses an auxiliary function gather_fanin_signals, that
+     * collects the fanin data for a node and matches them in the virtual_pi_network.
+     */
+    explicit virtual_pi_network(Ntk& ntk, std::vector<std::vector<mockturtle::node<Ntk>>>& ntk_lvls,
```
please put this one to the other constructors.

Drewniok replied:
This is a very long constructor. Can we break it into smaller pieces?

Drewniok commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,516 @@
+        {
+            std::vector<typename Ntk::signal> children{};
+            const auto                        edge_it_int      = edge_it;
+            int                               first_fi_edge_it = -1;
```
```suggestion
            int64_t                               first_fi_edge_it = -1;
```
?

Drewniok commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,516 @@
+                                  const auto fn           = ntk.get_node(f);
+                                  auto       tgt_signal_v = old2new_v[fn];
+
+                                  assert(edge_it_int < lvl.size());
```
Please always add an output string when using assert.

Drewniok commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,516 @@
+     */
+    signal create_virtual_pi(const signal& real_pi)
+    {
+        signal s = Ntk::create_pi();
```
```suggestion
        const signal s = Ntk::create_pi();
```

Drewniok commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,516 @@
+     */
+    [[nodiscard]] auto get_real_pi(const node& v_pi) const
+    {
+        auto it = _storage->map_virt_to_real_pi.find(v_pi);
```
```suggestion
        const auto it = _storage->map_virt_to_real_pi.find(v_pi);
```

Drewniok commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,516 @@
+    /**
+     * Shared pointer if the virtual PI storage.
+     */
+    std::shared_ptr<virtual_storage<Ntk>> _storage;
```
```suggestion
    storage strg;
```
if you add ``using storage = std::shared_ptr<virtual_storage<Ntk>>;`` in the first line after ``public``. This is how we do it in _fiction_.

Drewniok commented on include/fiction/networks/virtual_pi_network.hpp:
Please check which functions can be private or public.

Drewniok commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,516 @@
+        _storage->map_virt_to_real_pi.clear();
+    }
+
+  protected:
```
```suggestion
  private:
```
?

Drewniok reviewed (commented):
As discussed with @marcelwa and you (@hibenj), I have done a review of the files that have not yet been reviewed by Marcel to assist Marcel since we have many open PRs at the moment. 

I left a few tips and comments that Marcel taught me over time. So I would like to pass them on to you :)

I didn't go through the docstrings in detail because I thought it would be helpful to review the code first. Similar problems/issues/recommendations have only been addressed once. So please read all comments first to get a better understanding and overview. 

Thanks for the implementation, I really enjoyed the review so far and learned a lot. I hope my feedback helps in some way! 😄

hibenj commented on include/fiction/algorithms/network_transformation/node_duplication_planarization.hpp:
```diff
@@ -0,0 +1,468 @@
+     * The output order determines the starting layer for this algorithm. If this option is turned off, the output order
+     * remains the same as in the provided network. If it is turned on, the outputs are ordered randomly.
+     */
+    bool random_output_order = false;
```
I have seen most structs which hold _params to use bools. Also i am not sure how a bool should be replaced with an enum. could you provide an example?

Drewniok replied:
@hibenj Sure! No worries. You can take a look at ``write_operational_domain.hpp`` or ``quickexact.hpp``.

hibenj commented on include/fiction/algorithms/network_transformation/node_duplication_planarization.hpp:
```diff
@@ -0,0 +1,468 @@
+
+    virtual_pi_network<Ntk> run(std::vector<std::vector<mockturtle::node<Ntk>>>& ntk_lvls_new)
+    {
+        const bool border_pis = true;
```
i wanted to have a functionality that is not implemented yet, i marked it with todo

hibenj commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,693 @@
+    explicit virtual_pi_network(Ntk& ntk, std::vector<std::vector<mockturtle::node<Ntk>>>& ntk_lvls,
+                                std::vector<std::vector<mockturtle::node<Ntk>>>& ntk_lvls_new)
+    {
+        std::unordered_map<mockturtle::node<Ntk>, bool> node_status;
```
but it is only used for the PI creation. `mockturtle::node_map` is initialized with a full network, right?

hibenj commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,693 @@
+     *
+     */
+    template <typename Ntk>
+    explicit virtual_pi_network(Ntk& ntk, std::vector<std::vector<mockturtle::node<Ntk>>>& ntk_lvls,
+                                std::vector<std::vector<mockturtle::node<Ntk>>>& ntk_lvls_new)
```
This function takes an array of duplicated nodes as input. A level looks e.g. like this: `[2, 3, 2]`. This means 2 is duplicated. The function will create a new node for the second `2` and makes sure all fanin relations are planar.

hibenj commented on include/fiction/networks/virtual_pi_network.hpp:
```diff
@@ -0,0 +1,693 @@
+         * @param edge_it Iterator for edge.
+         * @return Vector of fanins in the virtual_pi_network connected to the processed node.
+         */
+        const auto gather_fanin_signals = [&](const auto& n, const auto& lvl, size_t& edge_it)
```
This is now part of `node_duplication_planarization.hpp`. This is all outdated i guess

hibenj commented on include/fiction/networks/views/extended_rank_view.hpp:
The issue is that i allow for a different construction of an extended_rank_view and also i allow easier modification of the private members. when inheriting from rank_view i am not able to modify the private members. DO you know a better solution for this?

Drewniok commented:
@hibenj No worries, the CI fails on Mac because you accidentally updated the submodules. You should revert those changes to fix this :)

hibenj commented:
Closed, because it is covered by other PRs
